src/core/ext/filters/http/message_compress/message_compress_filter.cc

```diff
@@ -366,6 +380,7 @@ static void start_send_message_batch(void* arg, grpc_error* unused) {
           calld->send_initial_metadata_state == HAS_COMPRESSION_ALGORITHM)) {
     send_message_batch_continue(elem);
   } else {
+    maybe_initialize(elem, calld);
```

Comment:
How about putting this at the place where `calld->send_initial_metadata_state` is set to `HAS_COMPRESSION_ALGORITHM`? (currently it is at line 426). In that case you probably can avoid the "maybe" term and `state_initialized`.

Reply:
> Is that common not to have compression algorithm in the initial metadata, and some default compression algorithm?

Ah that is good point! In fact, I am suspecting it is *probably* a bug in the `process_send_initial_metadata`; `has_compression_algorithm` probably should not set to true if the [compression algorithm](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/http/message_compress/message_compress_filter.cc#L166) parsed from the initial metadata is actually "no compression" (@AspirinSJL could you also take a look?). But you are totally right that under this situation, it should not initialize those variables.

Reply:
Yes I agree. We probably shouldn't has_compression to true in the line you mentioned.

Reply:
That seems a bug to me.. `has_compression_algorithm` is always set to true in that function. The code looks strange. I guess we need to stop processing when there is any parsing error.

Reply:
I'm not very familiar with that code, but I can try to change it. If no test fails, then let's change it.

Reply:
Trying to fix this in https://github.com/grpc/grpc/pull/19239.

Reply:
Thank you for the fix and sorry for the delay! 

I rebased and did what @muxi suggested.  PTAL. thanks!